### PR TITLE
Add server/client sync filters

### PR DESCRIPTION
**Pull request by Arlen22**
This adds the server and client sync filters as proposed by me in #5342. 

The changes are mainly to the syncer, and apply to the file system adapter as well, but that does not break anything as write-only tiddlers will still be saved, and read-only tiddlers are ignored anyway because the file system adapter does not read the file system. In the future, whatever mechanism is used for bags will most likely overcome this problem.

### CODE REVIEW

**Comment by Jermolene** [comment]
Thanks @Arlen22 I've had a quick look. What about backwards compatibility? That is the constraint that gave rise to the current implementation.

**Comment by Arlen22** [comment]
It retains compatibility with the current implementation because if nothing is specified in the client sync filter (which is appended to the server sync filter) then it functions exactly the same. The only difference is that I changed the sync filter to match what the tiddlyweb adapter requests. I should probably take a look at that and also make sure I didn't miss any other uses of the sync filter.

**Comment by Jermolene** [comment]
Hi @Arlen22 perhaps I'm reading the PR wrongly, but it seems that the syncer is no longer obeying `$:/config/SyncFilter`, which wouldn't be backwards compatible for users that have customised that tiddler.

File: plugins/tiddlywiki/tiddlyweb/config-tiddlers-filter.tid [hunk 1]
```diff
@@ -1,2 +1,2 @@
-title: $:/config/Server/ExternalFilters/[all[tiddlers]] -[[$:/isEncrypted]] -[prefix[$:/temp/]] -[prefix[$:/status/]] -[[$:/boot/boot.js]] -[[$:/boot/bootprefix.js]] -[[$:/library/sjcl.js]] -[[$:/core]]
+title: $:/config/Server/ExternalFilters/[is[tiddler]] -[[$:/core]] -[[$:/library/sjcl.js]] -[prefix[$:/boot/]] -[prefix[$:/HistoryList]] -[status[pending]plugin-type[import]] -[[$:/isEncrypted]] -[prefix[$:/status/]] -[prefix[$:/state/]] -[prefix[$:/temp/]]
```
Comment (pmario) [inline plugins/tiddlywiki/tiddlyweb/config-tiddlers-filter.tid:1]: IMO a filter should reduce the number of elements as fast as possible. I'm sure there are more `[prfix[xxx]]` elements than `$:/core` in the list. Why did you change the order in that way?

**Comment by Arlen22** [comment]
@Jermolene  and @pmario, The old version (red line) in the change @pmario commented on was what tiddlers.json was returning for the skinny tiddler list and the new version (green line) was what the `$:/config/SyncFilter` was set to (which I renamed `$:/config/SyncServerFilter`. In retrospect I think it would be easier at the moment to change the name back to SyncFilter because it is used elsewhere as well. 

So basically the only change I made here was that tiddlers.json using the SyncFilter instead of having its own separate filter that's hardcoded.

File: plugins/tiddlywiki/tiddlyweb/config-tiddlers-filter.tid (hunk 1 above)
Comment (Arlen22) [reply plugins/tiddlywiki/tiddlyweb/config-tiddlers-filter.tid:1]: See my comment here: https://github.com/Jermolene/TiddlyWiki5/pull/5384#issuecomment-758172726

**Comment by Arlen22** [comment]
The other change I made was to have a separate filter appended to the SyncFilter (which I named SyncClientFilter) to change the tiddlers that can be synced back to the server if desired. This is a more granular approach obviously than just excluding system tiddlers from tiddlers.json and also allows tiddlers to be made readonly by subtracting them in the filter. 

I explained it better at https://github.com/Jermolene/TiddlyWiki5/issues/5342#issuecomment-754251110 so I copied that here to try to explain it better. 

Basically, `SyncServerFilter` and `SyncClientFilter` work in tandom. `SyncServerFilter` would affect which tiddlers get synced to the client through all of the several sync mechanisms, and `SyncClientFilter` would affect which tiddlers get synced back to the server. I'm thinking that it would be chained to `SyncServerFilter` so it would add or remove tiddlers to either make tiddlers readonly from the server or write-only to the server. 

Here is an example of what I'm talking about. 

```txt
SyncServerFilter: [all[tiddler]] -[[$:/core]] -[[$:/StoryList]]
SyncClientFilter: [[$:/StoryList]] -[[$:/ServerStatus]] +[tag[Only Save Tiddlers Tagged With This!]]
```

I'm not sure exactly how that would get executed, but this is what I'm imagining. Additional tiddlers can be added or removed to `SyncClientFilter` as shown above. As you can see, adding a plus sign should modify everything prior to it in the filter, if I understand right. 

```js
let tiddlersToServe = `${SyncServerFilter}`;
let tiddlersToSave = `${SyncServerFilter} ${SyncClientFilter}`;
```

`tiddlersToServe` would be used to filter the list of tiddlers that goes from the server to the client (and also to filter the list of lazy-load requests), and `tiddlersToSave` would be used to filter which changes are saved from the client to the server. Obviously both the server and the client would have this check (the server in the PUT and DELETE tiddlers route and the client wherever needed).

**Comment by pmario** [comment]
I think this doesn't work. With the default settings the `$:/StoryList` has to be synced in both directions. 

For me it would be OK, If I have a "per user" "add-list", that allows my to tell the server which tiddlers I want to be shipped to the client. The "add-list" may be a second request. I don't have a problem with 2 requests 1 big list with content 1 small list with additional content

**Comment by Arlen22** [comment]
> I think this doesn't work. With the default settings the `$:/StoryList` has to be synced in both directions.

I believe this is correct and should still be the case. If someone doesn't want to sync $:/StoryList they would remove it from SyncFilter so it does not get sent from the server and then add it to SyncClientFilter so it does get saved back to the server.

**Comment by pmario** [comment]
@Arlen22 ... Is this PR still something we should consider?
